DR-1404: Billing Profile Update

snf2ye opened the pull request:
Main Change: Adding updateProfile endpoint
- Input: BillingProfileUpdateModel that includes profile id (UUID), billing account id, and description. 
- Async flight is kicked off that operates in 4 steps:
1. Locate existing billing profile by profile id and store in flight context to be used in case of "UNDO"
2. Update metadata: in profile resource db table, update the billing account and description
3. Verify that the user has access to the new billing account
4. Gather ids of all projects associated with the billing profile, and send a request to each google project to update the billing account

Notes:
- "No Spend" Billing Account: Adding a second billing account id to be used in connected tests. It should NOT be used for general testing because we do not want to spend funds on this billing account. It is added in order to test the ability to switch billing accounts successfully. 
- Fixed bug w/ ProfileRequestValidator: if the billing account id field was missing from the request model, it would fail with a 500 Internal server error instead of a "bad request" error.

snf2ye committed:
wip

snf2ye committed:
fully working workflow

snf2ye committed:
wip

snf2ye committed:
billing service connected test

snf2ye committed:
remove SAM resource update - not needed

snf2ye committed:
rework request model

snf2ye committed:
update IAM check

snf2ye committed:
clean up test

snf2ye committed:
removed wrong comment

snf2ye commented:
Note on test failures: Only one integration test failed (FileTest.fileUncommonNameTest), and it doesn't appear to be from my changes. ([scan here](fileUncommonNameTest)) I'll keep an eye on it as tests run after review changes.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileDao.java:
```diff
@@ -74,6 +75,28 @@ public BillingProfileModel createBillingProfile(BillingProfileRequestModel profi
+            .addValue("billing_account_id", profileRequest.getBillingAccountId())
+            .addValue("description", profileRequest.getDescription());
+        DaoKeyHolder keyHolder = new DaoKeyHolder();
+        jdbcTemplate.update(sql, params, keyHolder);
```
Do we normally not check results from `update()`? I would expect something like 
```
        int numRows = jdbcTemplate.update(sql, params, keyHolder);
        if (numRows != 1) {
             throw ...;
        }
```
Unless we expect this to modify more than one row?

nmalfroy replied:
I'm not sure that would be possible since id is a PK.  I think it has to be 0 or 1.  But it's a fair point that we may want to be more defensive...I do wonder if there's a way to do this via annotations so it's not quite as onerous to go through and do this since I don't believe we do these kinds of checks today

ddietterich replied:
We only do checks where the count is used for later logic. Not for the PK case

snf2ye replied:
I added a check here. As explained in the code comments, I'm throwing a "ProfileNotFoundException." I'd appreciate a sanity check on the assumptions I made here!

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -79,6 +81,29 @@ public String createProfile(BillingProfileRequestModel billingProfileRequest,
+    /**
+     * Update billing profile. We make the following checks:
+     * <ul>
+     *     <le>The service must have proper permissions on the google billing account</le>
```
Is this correct? I'm not familiar with `<le>`. I would expect `<li>` here.

ddietterich replied:
haha, yes it should be `<li>`. This is a cut'n'paste from a comment I wrote.
`<le>` is from a formatting tool I used a long time ago - "list element"

snf2ye replied:
Fixed!

pshapiro4broad commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -79,6 +81,29 @@ public String createProfile(BillingProfileRequestModel billingProfileRequest,
+
+        String description = String.format("Update billing for profile id '%s'", billingProfileRequest.getId());
+        return jobService
+            .newJob(description, ProfileUpdateFlight.class, billingProfileRequest, user)
```
This is probably an existing convention but I find it a bit confusing that this service both updates the actual data (by using the DAO) and creates flights, which in turn call this service to call the DAO.

I would expect a division where one class only creates flights (which wouldn't need a DAO) and one class only modifies the data.

ddietterich replied:
This service provides the complete implementation for profile operations. It is not a concern of classes calling this service whether it uses flights or not. This is consistent practice across the codebase. The *Service classes provide the complete API for their service. 

We have tried to eliminate direct references to dao classes from other service classes, and have other service classes and flights only use the service layer. This is not consistent; we instituted the practice but have not gone up and cleaned out all of the all usage. We have not done a complete job of making things packaged local. In fact because we use the `@Transaction` annotation, some dao methods need to be public so the annotation to work.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/flight/update/ProfileUpdateFlight.java:
```diff
@@ -0,0 +1,35 @@
+        GoogleProjectService googleProjectService =
+            (GoogleProjectService) appContext.getBean("googleProjectService");
+
+        bio.terra.model.BillingProfileUpdateModel request =
```
Why not import?

snf2ye replied:
Fixed!

pshapiro4broad commented on src/main/java/bio/terra/service/profile/flight/update/UpdateProfileMetadataStep.java:
```diff
@@ -0,0 +1,68 @@
+        FlightMap workingMap = context.getWorkingMap();
+        workingMap.put(JobMapKeys.REVERT_TO.getKeyName(), oldProfileModel);
+
+        // TODO: Is there a better way to handle this?
```
This seems OK to me but I would only check for `null`, unless you know there's a way an empty string can end up here and everything else is correct.

pshapiro4broad commented on src/main/java/bio/terra/service/profile/flight/update/UpdateProfileMetadataStep.java:
```diff
@@ -0,0 +1,68 @@
+        logger.info("Profile update failed. Reverting to old profile metadata.");
+        FlightMap workingMap = context.getWorkingMap();
+        BillingProfileModel oldProfileModel =
+            workingMap.get(JobMapKeys.REVERT_TO.getKeyName(), BillingProfileModel.class);
```
I'm not familiar with how stairway works. If `doStep()` fails during execution, does stairway run `undoStep()`? If so, I don't think we can depend on `REVERT_TO` here as it wouldn't have been persisted, since it's set within this step, not before it.

nmalfroy replied:
Yes, you might need to do a null check after this line.  Or be cute with optionals, with something like:
```
Optional.ofNullable(workingMap.get(JobMapKeys.REVERT_TO.getKeyName(), BillingProfileModel.class)).ifPresent(m -> {
    BillingProfileUpdateModel requestOldModel = new BillingProfileUpdateModel()
            .id(m.getId())
            .billingAccountId(m.getBillingAccountId())
            .description(m.getDescription());
    profileService.updateProfileMetadata(requestOldModel);
});
```

ddietterich replied:
I think the cleanest way to do this is to have an initial step that sets up the REVERT_TO.

You will never catch me using an Optional and lambda to say `if (x == null)`
I remain unafraid of null! :)

snf2ye replied:
Ok - I created a new step to define the REVERT_TO object.

pshapiro4broad reviewed (commented):
just a few comments

nmalfroy commented on src/main/java/bio/terra/service/profile/flight/update/UpdateProfileMetadataStep.java:
```diff
@@ -0,0 +1,68 @@
+        // TODO: Is there a better way to handle this?
+        // I want to allow for just updating a description or just updating the billing account
+        // build request - handle if description or billing account id is not included
+        if (profileRequest.getBillingAccountId() == null || profileRequest.getBillingAccountId().isEmpty()) {
```
I think you can do this more simply with `StringUtils.isEmpty(profileRequest.getBillingAccountId())` from `org.apache.commons.lang3.StringUtils`...This does the null check for you

nmalfroy replied:
Actually, comparing with phil's comment above, I agree that we may want to only check for nulls since someone could conceivably want to set the accountId or description to empty.  That'd be the only way to unset the billing id (which seems useful as a way to disable billing, perhaps)

ddietterich replied:
We shouldn't allow billing to be unset. Billing accounts can be disabled, if that is desired, but we need to keep the link between billing profile, project, and the dataset, snapshot, and bucket objects using it.

Is it a REST API convention that leaving something null in a PUT means don't change it? Maybe a more explicit directive in the request would be better?

Do we require that something is specified to be updated?

snf2ye replied:
Ok - I made a change to require both description and billing account id in the request. My thoughts on this decision are described in one of the other comments: https://github.com/DataBiosphere/jade-data-repo/pull/747#discussion_r542521663

nmalfroy commented on src/main/java/bio/terra/service/profile/flight/update/UpdateProfileUpdateGCloudProject.java:
```diff
@@ -0,0 +1,50 @@
+
+    @Override
+    public StepResult doStep(FlightContext context) throws InterruptedException {
+        // TODO: Should I check if billing account is different from request?
```
I agree.  This is a seldom used operation so optimizing for minimize DB writes I think isn't worth the headaches we may eventually get if a thing doesn't update when we really want it to

snf2ye replied:
Awesome - leaving code as is and removing comment.

nmalfroy commented on src/main/resources/api/data-repository-openapi.yaml:
```diff
@@ -2380,6 +2435,19 @@ components:
+          $ref: '#/components/schemas/UniqueIdProperty'
+        billingAccountId:
+          type: string
+          description: Unique identifier of the billing account from Google
```
I forget, is this a UUID?  If so, we maybe want to mention that it is (though I'm not 100% sure that it is).  We may want to be explicit since I believe that google uses two ids...

ddietterich replied:
No, it is not a UUID. According to Jeremy's comment in the code:
`Google billing account ids are three sets of six capitalized hex characters separated by dashes`

I spent a few minutes trying to confirm that, but didn't find it spelled out.

fboulnois replied:
Yes, that is correct. It looks like `012345-6789AB-CDEF00` .

nmalfroy commented on src/test/java/bio/terra/app/configuration/ConnectedTestConfiguration.java:
```diff
@@ -36,4 +37,13 @@ public String getGoogleBillingAccountId() {
     }
+
+    // TODO- IS THERE A BETTER NAME FOR THIS?
+    public String getSecondGoogleBillingAccountId() {
```
This name isn't too bad.  We could theoretically change billingId to an array, but that would be a huge pain with little payoff.  Maybe alternate if we never expect it to be greater than two billing Ids in our tests?

snf2ye replied:
Per Dan's suggestion, I changed it to "noSpendGoogleBillingAccountId".

nmalfroy commented on src/test/java/bio/terra/service/profile/GoogleBillingServiceTest.java:
```diff
@@ -0,0 +1,143 @@
+
+    // Ignoring since this manipulates billing accounts
+    // TODO add back in when we can create a new project, test changing billing accounts, and then delete project
+    @Ignore
```
Is it possible to add a comment directly in the annotation? I think it then shows up in the logs. Not 100% but might be worth it

ddietterich replied:
It does appear that `@Ignore` takes a string. Nice. So many fun little things to learn

snf2ye replied:
done!

nmalfroy commented on src/test/java/bio/terra/service/profile/GoogleBillingServiceTest.java:
```diff
@@ -0,0 +1,143 @@
+        newBillingProfile.setBillingAccountId(newBillingAccountId);
+        boolean billingEnabled = googleBillingService.assignProjectBilling(newBillingProfile, projectResource);
+        resetBillingAccount = true;
+        assertTrue("Billing should be enabled after updating the billing account", billingEnabled);
```
At some point we should agree on assertJ vs. JUnit to assertions.  We've got a mix of the two

ddietterich replied:
A good time to do that would be when we switch to JUnit 5. It makes some incompatible changes to the JUnit assertions.

nmalfroy commented on src/test/java/bio/terra/service/profile/ProfileServiceTest.java:
```diff
@@ -0,0 +1,130 @@
+
+        BillingProfileModel newModel = connectedOperations.updateProfile(updatedRequest);
+        logger.info("Updated model: {}", newModel.toString());
+        assertThat("AFTER UPDATEBilling account should be equal to the newBillingAccountId",
```
Should this be: `AFTER UPDATE: Billing...`?

nmalfroy commented on src/test/java/bio/terra/service/profile/ProfileServiceTest.java:
```diff
@@ -0,0 +1,130 @@
+    @Ignore
+    @Test
+    public void updateProfileTest() throws Exception {
+        System.out.println("profile: " + profile.getProfileName());
```
I'd use a logger for this class for printing out things.  This way it shows up inline with exceptions etc.

nmalfroy reviewed (changes requested):
This is looking good!  A couple of things around allowing clearing billing Ids and descriptions and a couple of nits

ddietterich commented on src/main/java/bio/terra/service/profile/flight/update/UpdateProfileMetadataStep.java:
```diff
@@ -0,0 +1,68 @@
+        // Update to new billing metadata
+        BillingProfileModel profileModel = profileService.updateProfileMetadata(profileRequest);
+        workingMap.put(JobMapKeys.RESPONSE.getKeyName(), profileModel);
+        workingMap.put(JobMapKeys.STATUS_CODE.getKeyName(), HttpStatus.CREATED);
```
I think you want OK for the status here, not CREATED.

snf2ye replied:
Thanks for catching that!

ddietterich commented on src/main/java/bio/terra/service/profile/flight/update/ProfileUpdateFlight.java:
```diff
@@ -0,0 +1,35 @@
+        //update billing account id metadata
+        addStep(new UpdateProfileMetadataStep(profileService, request, user));
+        // Make sure valid account before changing in gcloud project
+        addStep(new UpdateProfileVerifyAccountStep(profileService, user));
```
You can determine here what the flight is updating. If it is only the description, only generate the metadata step.
If it is the billing account, then generate all three steps.

You could pass that into the steps, if they need to know, rather than having them each figure it out.

snf2ye replied:
That would make a lot more sense! 

I decided to simplify my solution by requiring both a billingAccountId and description. I decided to do this for a few reasons:
- By requiring both fields, the user experience will be clear: The billing account and description will be updated according to their input. 
- There is a workaround if only one needs to be updated: The user can "retrieve" the profile, and set the billing account or description to the current value. 
- If billing account is updated, then most likely an update to the description would be useful. However, I think there may be a case for just updating the description. 
- This avoids early over-optimization. By requiring both fields, the solution is simpler, but the full functionality can be achieved. We can always customize this in the future to better fit the needs of the user.

ddietterich commented on src/main/java/bio/terra/service/resourcemanagement/google/GoogleProjectService.java:
```diff
@@ -461,4 +461,34 @@ private ServiceUsage serviceUsage() throws IOException, GeneralSecurityException
+                .googleProjectNumber(googleProjectNumber);
+
+        // The billing profile has already been authorized so we do no further checking here
+        billingService.assignProjectBilling(billingProfile, googleProjectResource);
```
Do we know anything about possible errors here or the need for retries? (We would probably do it in billingService, but thought of it here...)

ddietterich commented on src/test/resources/application-connectedtest.properties:
```diff
@@ -2,6 +2,7 @@
 ct.ingestBucket=jade-testdata
 ct.ingestRequesterPaysBucket=jade_testbucket_requester_pays
 ct.googleBillingAccountId=00708C-45D19D-27AAFA
+ct.secondGoogleBillingAccountId=01A82E-CA8A14-367457
```
This is the billing account that is only supposed to be used in billing account testing, right?

Maybe add some documentation here? I wouldn't want someone to naively spend a bunch of money against that account. Maybe we should call it the "noSpend" account or some such.

snf2ye replied:
Oh, I like that idea! I've rename the variable and added documentation.

ddietterich reviewed (changes requested):
Two suggestions for the flight and a few questions.

snf2ye committed:
update from le to li

snf2ye committed:
import the model

snf2ye committed:
require billing account and description

snf2ye committed:
move retrieve existing to new step; rename to "nospend" billing account

snf2ye committed:
Validate update request

snf2ye committed:
update ignore call

snf2ye committed:
Add check for failed update and unit tests

snf2ye commented:
Thank you all for the reviews! I think I've addressed all of the comments. I'm going take another look and make sure I haven't broken any of the tests, and then will bug people for another round of reviews.

snf2ye committed:
Merge branch 'develop' into sh-DR-1404-profile-update

snf2ye committed:
fix checkstyles

snf2ye committed:
switch to no Id check for getProfileById

snf2ye committed:
Fix validation for create profile and update profile; add associated tests

snf2ye committed:
fix checkstyles

ddietterich reviewed (approved):
lgtm

nmalfroy commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -60,9 +62,9 @@ public ProfileService(ProfileDao profileDao,
-     *     <le>The service must have proper permissions on the google billing account</le>
-     *     <le>The caller must have billing.resourceAssociation.create permission on the google billing account</le>
-     *     <le>The google billing account must be enabled</le>
+     *     <li>The service must have proper permissions on the google billing account</le>
```
You still have the `</le>` closing tags

snf2ye replied:
Oh, good catch - fixing now.

nmalfroy commented on src/main/java/bio/terra/service/profile/ProfileService.java:
```diff
@@ -79,11 +81,34 @@ public String createProfile(BillingProfileRequestModel billingProfileRequest,
      * <ul>
-     *     <le>the caller must be an owner of the billing profile</le>
-     *     <le>There must be no dependencies on the billing profile;
+     *     <li>the caller must be an owner of the billing profile</le>
```
same thing with the closing tags

nmalfroy reviewed (approved):
Looks good!  Mind just fixing up the javadoc?

snf2ye committed:
Merge branch 'develop' into sh-DR-1404-profile-update

snf2ye committed:
switch to li

snf2ye committed:
switch back to just passing billing account id to verifyAccount

snf2ye commented:
Test run:
* One test failure in connected tests https://scans.gradle.com/s/yafzvdht77via
* One test failure for integration tests:https://scans.gradle.com/s/ah2pqkjd3fzwo/tests/:testIntegration/bio.terra.service.snapshot.SnapshotTest/snapshotUnauthorizedPermissionsTest#1

I'm going to kick tests off again.

snf2ye commented:
Re-run of tests:
* Connected tests passed
* Int test failed with 3 different test failures (Not the same 3 tests as failed in last run) https://scans.gradle.com/s/rwxcgcjayh2ri
This points to test flakiness.
